File: 프로그래머스/4/1843. 사칙연산/사칙연산.py
INF = 987654321
def solution(arr):
    
    op = []
    number = []
    for i in arr:
        try:
            number.append(int(i))
        except:
            op.append(i)
            
    n = len(arr) // 2 + 1
    MAX = [[-INF] * n for _ in range(n)]
    MIN = [[INF] * n for _ in range(n)]
    for step in range(n):
        for i in range(n-step):
            j = i + step
            if step == 0:
                MAX[i][j] = number[i]
                MIN[i][j] = number[i]
            for k in range(i, j):
                if op[k] == "+":
                    MAX[i][j] = max(MAX[i][j], MAX[i][k] + MAX[k+1][j])
                    MIN[i][j] = min(MIN[i][j], MIN[i][k] + MIN[k+1][j])
                else:
                    MAX[i][j] = max(MAX[i][j], MAX[i][k] - MIN[k+1][j])
                    MIN[i][j] = min(MIN[i][j], MIN[i][k] - MAX[k+1][j])
    return MAX[0][n-1]

File: 프로그래머스/4/1843. 사칙연산/사칙연산.java
import java.util.List;
import java.util.ArrayList;

class Solution {
    public int solution(String arr[]) {
        int answer = -1;
        
        int[] numbers = new int[arr.length/2+1];
        String[] op = new String[arr.length/2];
        int idx = 0;
        int idx2 = 0;
        for (String s : arr) {
            if (Character.isDigit(s.charAt(0))) {
                numbers[idx++] = Integer.parseInt(s);
            } else {
                op[idx2++] = s;
            }
        }
        
        int[][] max = new int[arr.length/2+1][arr.length/2+1];
        int[][] min = new int[arr.length/2+1][arr.length/2+1];
       
        for (int i=0; i<arr.length/2+1; i++) {
            for (int j=0; j<arr.length/2+1; j++) {
                max[i][j] = Integer.MIN_VALUE;
                min[i][j] = Integer.MAX_VALUE;
            }
        }
        for (int step = 0; step < arr.length/2+1; step++){
            for(int i =0; i < arr.length/2+1 - step; i++){
                int j = i + step;
                if (step == 0) {
                    min[i][j] = numbers[i];
                    max[i][j] = numbers[i];
                }
                
                for (int k = i; k < j; k++) {
                    
                    if (op[k].equals("+")) {
                        max[i][j] = Math.max(max[i][j], max[i][k] + max[k+1][j]);
                        min[i][j] = Math.min(min[i][j], min[i][k] + min[k+1][j]);
                    } else if(op[k].equals("-")) {
                        max[i][j] = Math.max(max[i][j], max[i][k] - min[k+1][j]);
                        min[i][j] = Math.min(min[i][j], min[i][k] - max[k+1][j]);
                    }
                }
            }
        }
        return max[0][arr.length/2];
    }
}